### ISSUE TITLE
[BUG] Groupby aggregations on multiple columns fail with second column not found

### ISSUE THREAD

**Issue by beckernick** [rapidsai/dask-cudf#125]
## Description
As a user, I want to compute hierarchical aggregations by grouping on multiple columns. Passing a list of column names to the `groupby` API causes a column not found error, while grouping by each column individually succeeds. The second column in the list is the one that isn't found.

## Standard Dask Example (Expected Result)
```
import cudf
import numpy as np
import pandas as pd
import dask_cudf
import dask.dataframe as dd

df = cudf.DataFrame([('a', list(range(20))),
('b', list(reversed(range(20)))),
('c', list(range(20)))])

df['agg_col1'] = [1 if x % 2 == 0 else 0 for x in range(len(df))]
df['agg_col2'] = [1 if x % 3 == 0 else 0 for x in range(len(df))]

pdf = df.to_pandas()
pddf = dd.from_pandas(pdf, npartitions=2)

pddf.groupby(['agg_col1', 'agg_col2']).sum().compute()
                    a   b   c
agg_col1 agg_col2            
0        0         73  60  73
         1         27  30  27
1        0         54  60  54
         1         36  40  36
```

## Dask Cudf Example
```
import cudf
import numpy as np
import pandas as pd
import dask_cudf
import dask.dataframe as dd

df = cudf.DataFrame([('a', list(range(20))),
('b', list(reversed(range(20)))),
('c', list(range(20)))])

df['agg_col1'] = [1 if x % 2 == 0 else 0 for x in range(len(df))]
df['agg_col2'] = [1 if x % 3 == 0 else 0 for x in range(len(df))]

ddf = dask_cudf.from_cudf(df, npartitions=2)

print(ddf.groupby('agg_col1').sum().compute())                                                                                    
     a    b    c  agg_col2
0  100   90  100         3
1   90  100   90         4

print(ddf.groupby('agg_col2').sum().compute())                                                                                    
     a    b    c  agg_col1
0  127  120  127         6
1   63   70   63         4

print(ddf.groupby(['agg_col1', 'agg_col2']).sum().compute())
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-61-8baea99ae3bc> in <module>
     14 ddf = dask_cudf.from_cudf(df, npartitions=2)
     15 
---> 16 print(ddf.groupby(['agg_col1', 'agg_col2']).sum().compute())

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/base.py in compute(self, **kwargs)
    154         dask.base.compute
    155         """
--> 156         (result,) = compute(self, traverse=False, **kwargs)
    157         return result
    158 

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    396     keys = [x.__dask_keys__() for x in collections]
    397     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 398     results = schedule(dsk, keys, **kwargs)
    399     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    400 

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/local.py in get_sync(dsk, keys, **kwargs)
    499     """
    500     kwargs.pop('num_workers', None)    # if num_workers present, remove it
--> 501     return get_async(apply_sync, 1, dsk, keys, **kwargs)
    502 
    503 

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    445             # Seed initial tasks into the thread pool
    446             while state['ready'] and len(state['running']) < num_workers:
--> 447                 fire_task()
    448 
    449             # Main loop, wait on tasks to finish, insert new ones

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/local.py in fire_task()
    441                             args=(key, dumps((dsk[key], data)),
    442                                   dumps, loads, get_id, pack_exception),
--> 443                             callback=queue.put)
    444 
    445             # Seed initial tasks into the thread pool

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/local.py in apply_sync(func, args, kwds, callback)
    488 def apply_sync(func, args=(), kwds={}, callback=None):
    489     """ A naive synchronous version of apply_async """
--> 490     res = func(*args, **kwds)
    491     if callback is not None:
    492         callback(res)

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    233         failed = False
    234     except BaseException as e:
--> 235         result = pack_exception(e, dumps)
    236         failed = True
    237     return key, result, failed

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    228     try:
    229         task, data = loads(task_info)
--> 230         result = _execute_task(task, data)
    231         id = get_id()
    232         result = dumps((result, id))

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/compatibility.py in apply(func, args, kwargs)
     91     def apply(func, args, kwargs=None):
     92         if kwargs:
---> 93             return func(*args, **kwargs)
     94         else:
     95             return func(*args)

/conda/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/groupby.py in _apply_chunk(df, *index, **kwargs)
    235         if isinstance(columns, (tuple, list, set, pd.Index)):
    236             columns = list(columns)
--> 237         return func(g[columns])
    238 
    239 

/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf-0.6.0.dev0+803.gde773d4-py3.7-linux-x86_64.egg/cudf/groupby/groupby.py in __getitem__(self, arg)
    286             for val in arg:
    287                 if val not in self._val_columns:
--> 288                     raise KeyError("Column not found: " + str(val))
    289         result = self.copy()
    290         result._val_columns = arg

KeyError: 'Column not found: agg_col2'
```

**Comment by mrocklin** [comment]
Yup, currently cudf doesn't support MultiIndexes.  It's not clear what the
medium term solution to this is likely to be.

**Comment by beckernick** [comment]
Looks like you're right. We seem to have created a workaround in cuDF leveraging the GenericIndex and a semi-hacky integration of one of the groupby columns into the columns of the dataframe.

```
import cudf
import numpy as np
import pandas as pd
import dask_cudf
import dask.dataframe as dd

df = cudf.DataFrame([('a', list(range(20))),
('b', list(reversed(range(20)))),
('c', list(range(20)))])

df['agg_col1'] = [1 if x % 2 == 0 else 0 for x in range(len(df))]
df['agg_col2'] = [1 if x % 3 == 0 else 0 for x in range(len(df))]

out = df.groupby(['agg_col1', 'agg_col2']).sum()
print(out)
   agg_col2   a   b   c
agg_col1
0         0  73  60  73
0         1  27  30  27
1         0  54  60  54
1         1  36  40  36

out.index
GenericIndex([0, 0, 1, 1], dtype=int64)

out.columns
Index(['agg_col2', 'a', 'b', 'c'], dtype='object')
```

**Comment by mrocklin** [comment]
Failing tests in https://github.com/rapidsai/dask-cudf/pull/134

**Comment by quasiben** [comment]
After a bit of research within dask/cudf/dask-cudf I don't think we can support multicolumn groupby without MultiIndex support from cudf (https://github.com/rapidsai/cudf/issues/483) .  I believe dask chunks the groupby objects with a MultiIndex created from an internal call within dask based on the by parameter:

```python
self.index = _normalize_index(df, by)
```

I don't see how we can easily work around this in either the dask or cudf codebase and thus we need to wait for MultiIndex support from cudf

**Comment by beckernick** [comment]
When I run this same example now, I get a different error:

```
import cudf
import numpy as np
import pandas as pd
import dask_cudf
import dask.dataframe as dd

df = cudf.DataFrame([('a', list(range(20))),
('b', list(reversed(range(20)))),
('c', list(range(20)))])

df['agg_col1'] = [1 if x % 2 == 0 else 0 for x in range(len(df))]
df['agg_col2'] = [1 if x % 3 == 0 else 0 for x in range(len(df))]

ddf = dask_cudf.from_cudf(df, npartitions=2)
print(ddf.groupby(['agg_col1', 'agg_col2']).a.sum().compute())
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-275-93b4d7e62d4c> in <module>
     13 
     14 ddf = dask_cudf.from_cudf(df, npartitions=2)
---> 15 print(ddf.groupby(['agg_col1', 'agg_col2']).a.sum().compute())

/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py in compute(self, **kwargs)
    154         dask.base.compute
    155         """
--> 156         (result,) = compute(self, traverse=False, **kwargs)
    157         return result
    158 

/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    396     keys = [x.__dask_keys__() for x in collections]
    397     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 398     results = schedule(dsk, keys, **kwargs)
    399     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    400 

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in get_sync(dsk, keys, **kwargs)
    501     """
    502     kwargs.pop('num_workers', None)    # if num_workers present, remove it
--> 503     return get_async(apply_sync, 1, dsk, keys, **kwargs)
    504 
    505 

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    468 
    469                 while state['ready'] and len(state['running']) < num_workers:
--> 470                     fire_task()
    471 
    472             succeeded = True

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in fire_task()
    443                             args=(key, dumps((dsk[key], data)),
    444                                   dumps, loads, get_id, pack_exception),
--> 445                             callback=queue.put)
    446 
    447             # Seed initial tasks into the thread pool

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in apply_sync(func, args, kwds, callback)
    490 def apply_sync(func, args=(), kwds={}, callback=None):
    491     """ A naive synchronous version of apply_async """
--> 492     res = func(*args, **kwds)
    493     if callback is not None:
    494         callback(res)

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    233         failed = False
    234     except BaseException as e:
--> 235         result = pack_exception(e, dumps)
    236         failed = True
    237     return key, result, failed

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    228     try:
    229         task, data = loads(task_info)
--> 230         result = _execute_task(task, data)
    231         id = get_id()
    232         result = dumps((result, id))

/conda/envs/rapids/lib/python3.7/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    116     elif istask(arg):
    117         func, args = arg[0], arg[1:]
--> 118         args2 = [_execute_task(a, cache) for a in args]
    119         return func(*args2)
    120     elif not ishashable(arg):

/conda/envs/rapids/lib/python3.7/site-packages/dask/core.py in <listcomp>(.0)
    116     elif istask(arg):
    117         func, args = arg[0], arg[1:]
--> 118         args2 = [_execute_task(a, cache) for a in args]
    119         return func(*args2)
    120     elif not ishashable(arg):

/conda/envs/rapids/lib/python3.7/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    113     """
    114     if isinstance(arg, list):
--> 115         return [_execute_task(a, cache) for a in arg]
    116     elif istask(arg):
    117         func, args = arg[0], arg[1:]

/conda/envs/rapids/lib/python3.7/site-packages/dask/core.py in <listcomp>(.0)
    113     """
    114     if isinstance(arg, list):
--> 115         return [_execute_task(a, cache) for a in arg]
    116     elif istask(arg):
    117         func, args = arg[0], arg[1:]

/conda/envs/rapids/lib/python3.7/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/core.py in _concat(args)
     66     # this seems easier. TODO: don't do this.
     67     args2 = [i for i in args if len(i)]
---> 68     return args[0] if not args2 else methods.concat(args2, uniform=True)
     69 
     70 

/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/methods.py in concat(dfs, axis, join, uniform, filter_warning)
    263         func = concat_dispatch.dispatch(type(dfs[0]))
    264         return func(dfs, axis=axis, join=join, uniform=uniform,
--> 265                     filter_warning=filter_warning)
    266 
    267 

/conda/envs/rapids/lib/python3.7/site-packages/dask_cudf-0.0.0.dev0-py3.7.egg/dask_cudf/backends.py in concat_cudf(dfs, axis, join, uniform, filter_warning)
     31     assert axis == 0
     32     assert join == "outer"
---> 33     return cudf.concat(dfs)

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0+212.gbf3764e7.dirty-py3.7-linux-x86_64.egg/cudf/multi.py in concat(objs, axis, ignore_index)
     50 
     51     if typ is DataFrame:
---> 52         return DataFrame._concat(objs, axis=axis, ignore_index=ignore_index)
     53     elif typ is Series:
     54         return Series._concat(objs, axis=axis)

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0+212.gbf3764e7.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/dataframe.py in _concat(cls, objs, axis, ignore_index)
   1037             index = RangeIndex(sum(map(len, objs)))
   1038         else:
-> 1039             index = Index._concat([o.index for o in objs])
   1040         data = [(c, Series._concat([o[c] for o in objs], index=index))
   1041                 for c in objs[0].columns]

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0+212.gbf3764e7.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/index.py in _concat(cls, objs)
    125     @classmethod
    126     def _concat(cls, objs):
--> 127         data = Column._concat([o.as_column() for o in objs])
    128         return as_index(data)
    129 

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0+212.gbf3764e7.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/index.py in <listcomp>(.0)
    125     @classmethod
    126     def _concat(cls, objs):
--> 127         data = Column._concat([o.as_column() for o in objs])
    128         return as_index(data)
    129 

AttributeError: 'MultiIndex' object has no attribute 'as_column'
```

Because `MultiIndex` is a subclass of `Index` and the `as_column` method isn't implemented at the `Index` level but at the `RangeIndex` and `GenericIndex` level, we get this attribute error. We likely need to implement `as_column` for MultiIndex. I will raise an issue on cuDF.


I'm built from:

*dask-cudf*
commit 4ba5f93d8071429cfcd8f899abd936b3f85bfb3e (HEAD -> branch-0.8, origin/branch-0.8, origin/HEAD)
Author: Matthew Rocklin <mrocklin@gmail.com>
Date:   Fri May 10 21:45:26 2019 -0500

*cudf*
commit bf3764e7ce248fe08aa204a1f971ddd20f50eba3 (HEAD -> branch-0.8, origin/branch-0.8, origin/HEAD)
Merge: a451d99b 59a03101
Author: Mark Harris <mharris@nvidia.com>
Date:   Mon May 13 13:05:50 2019 +1000

**Comment by quasiben** [comment]
Which version of dask are you using ?

**Comment by thomcom** [comment]
It's possible it is a new issue - but yesterday I wrote code that prevents MultiColumn from trying to call as_column - instead, each column of the MultiIndex gets added to a groubpy dataframe.

**Comment by beckernick** [comment]
I still get this error with dask-cudf.

cudf:
```commit 065eee11ebf3d709b156b39df86b5eb8a4be2602 (HEAD -> branch-0.8, origin/branch-0.8, origin/HEAD)
Merge: 20c87a94 3b6d4401
Author: Mark Harris <mharris@nvidia.com>
Date:   Tue May 14 17:13:53 2019 +1000```

dask-cudf:
```
commit 4ba5f93d8071429cfcd8f899abd936b3f85bfb3e (HEAD -> branch-0.8, origin/branch-0.8, origin/HEAD)
Author: Matthew Rocklin <mrocklin@gmail.com>
Date:   Fri May 10 21:45:26 2019 -0500
```

dask:
```
root@38105919cd59:/dask-cudf# conda list | grep dask
dask-core                 1.2.2                      py_0    conda-forge
dask-cudf                 0.0.0.dev0               pypi_0    pypi
```

@quasiben , @thomcom, are either of you able to reproduce this?

**Comment by thomcom** [comment]
This is caused by `npartitions=2` - it works great with `npartitions=1`.

**Comment by thomcom** [comment]
I have a fix for this I'm working into a PR now.

**Comment by thomcom** [comment]
Please confirm this is fixed when you can @beckernick

**Comment by beckernick** [comment]
@thomcom, will build from the hotfix to test this, as your merged fix isn't yet in branch-0.8

**Comment by beckernick** [comment]
Resolved by rapidsai/cudf#1741. Thanks @thomcom . Closing.

**Comment by randerzander** [comment]
@beckernick @thomcom hmm.. looks like we have a new problem.

I just installed conda nightlies and with Nick's repro, I get:
```
>>> import cudf
>>> import numpy as np
>>> import pandas as pd
>>> import dask_cudf
>>> import dask.dataframe as dd
>>> 
>>> df = cudf.DataFrame([('a', list(range(20))),
... ('b', list(reversed(range(20)))),
... ('c', list(range(20)))])
>>> 
>>> df['agg_col1'] = [1 if x % 2 == 0 else 0 for x in range(len(df))]
>>> df['agg_col2'] = [1 if x % 3 == 0 else 0 for x in range(len(df))]
>>> 
>>> ddf = dask_cudf.from_cudf(df, npartitions=2)
>>> print(ddf.groupby(['agg_col1', 'agg_col2']).a.sum().compute())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/groupby.py", line 940, in sum
    split_out=split_out)
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/groupby.py", line 845, in _aca_agg
    split_out=split_out, split_out_setup=split_out_on_index)
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/core.py", line 3726, in apply_concat_apply
    if dfs else None))
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask/utils.py", line 417, in __call__
    return meth(arg, *args, **kwargs)
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask_cudf/backends.py", line 21, in make_meta_cudf
    return x.head(0)
  File "/conda/envs/rapids/lib/python3.6/site-packages/cudf/dataframe/dataframe.py", line 351, in head
    return self.iloc[:n]
  File "/conda/envs/rapids/lib/python3.6/site-packages/cudf/indexing.py", line 76, in __getitem__
    return self._getitem_multiindex_arg(arg)
AttributeError: '_DataFrameIlocIndexer' object has no attribute '_getitem_multiindex_arg'
```

**Comment by beckernick** [comment]
Confirmed that I run into the same issue @randerzander .

**Comment by quasiben** [comment]
@thomcom want to pair on this and look the issue on Tuesday ?

**Comment by thomcom** [comment]
Sure, the issue has been introduced by https://github.com/rapidsai/cudf/pull/1622 so maybe we should pull in @shwina also.

**Comment by thomcom** [comment]
PR for this pending.

**Comment by quasiben** [comment]
@beckernick I just tested with @thomcom recent PR into branch-0.8 everything is working.  Can you confirm and close ?

**Comment by randerzander** [comment]
With the caveat that you now need to append .to_pandas() on the compute result, this looks resolved!

Thanks @quasiben and @thomcom !
